Inconsistent path to item in transitive dependency crate in diagnostic when the re-export is `#[doc(hidden)]`

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

When referencing items in transitive dependency crates, the current diagnostic behavior prefers paths through a direct dependency crate, even if the path involves a `#[doc(hidden)]` re-export.

*In the examples below, `direct_dep::__private` is a `#[doc(hidden)]` module re-exporting items from `transitive_dep`, which is not an extern crate of the current crate. See the full code of the three crates at the bottom of the issue description.*

```
error[E0308]: mismatched types
  --> $DIR/hidden-reexport-of-transitive-dep-item.rs:10:44
   |
LL |     let _: direct_dep::__private::Struct = Struct;
   |            -----------------------------   ^^^^^^ expected `direct_dep::__private::Struct`, found `Struct`
   |            |
   |            expected due to this
   |
   = note: `Struct` and `direct_dep::__private::Struct` have similar names, but are actually distinct types
note: `Struct` is defined in the current crate
  --> $DIR/hidden-reexport-of-transitive-dep-item.rs:7:1
   |
LL | struct Struct;
   | ^^^^^^^^^^^^^
note: `direct_dep::__private::Struct` is defined in crate `transitive_dep`
  --> $DIR/auxiliary/transitive-dep.rs:3:1
   |
LL | pub struct Struct;
   | ^^^^^^^^^^^^^^^^^
```

Based on this, for consistent behavior, the same preference would be expected if the re-exported item is nested in a module in the re-exporting direct dependency crate. However, currently, rustc instead shows an inaccessible path through the transitive dependency crate, falling back to the bare definition path of the item. See the examples below (diagnostic notes removed for brevity).

Actual:

```
error[E0308]: mismatched types
  --> $DIR/hidden-reexport-of-nested-transitive-dep-item.rs:10:51
   |
LL |     let _: direct_dep::__private::inner::Struct = Struct;
   |            ------------------------------------   ^^^^^^ expected `transitive_dep::Struct`, found `Struct`
   |            |
   |            expected due to this
```

Expected:

```
error[E0308]: mismatched types
  --> $DIR/hidden-reexport-of-nested-transitive-dep-item.rs:10:51
   |
LL |     let _: direct_dep::__private::inner::Struct = Struct;
   |            ------------------------------------   ^^^^^^ expected `direct_dep::__private::inner::Struct`, found `Struct`
   |            |
   |            expected due to this
```

Since the `#[doc(hidden)]` (but actually accessible) path is already preferred in the more basic case, it should be the same in the more complicated cases. Note that the issue disappears entirely and the accessible path through the direct dependency crate is used in both cases if the `#[doc(hidden)]` attribute is removed from the re-exporting module.

Related PRs/issues:
* #87349
* #153477

---

Code to reproduce the diagnostics:

I tried to reduce this as much as possible to the point that changing anything about the example does not trigger the inconsistent diagnostic anymore.

*transitive-dep.rs*:
```rs
#![crate_type = "lib"]

pub struct Struct;
```

*direct-dep.rs*:
```rs
#![crate_type = "lib"]

extern crate transitive_dep;

mod private {
    // NOTE: Wrapping the re-export in a module is required to trigger the def path fallback,
    //       and removing it results in the intended diagnostic in the first example
    //       with the accessible `#[doc(hidden)]` path through `direct_dep`.
    pub mod inner {
        pub use crate::transitive_dep::Struct;
    }
}

#[doc(hidden)]
pub mod __private {
    pub use crate::private::*;
}
```

*test.rs*:
```rs
extern crate direct_dep;

struct Struct;

fn main() {
    let _: direct_dep::__private::inner::Struct = Struct;
}
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.99.0-dev
binary: rustc
commit-hash: ac68faa20c58cbccd01ee7208bf3b6e93a7d7f96
commit-date: 2026-07-24
host: x86_64-unknown-linux-gnu
release: 1.99.0-dev
LLVM version: 22.1.8
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

--- a/compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs
+++ b/compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs
@@ -501,14 +501,14 @@ pub(in crate::rmeta) fn provide(providers: &mut Providers) {
                 }
 
                 if let Some(def_id) = child.res.opt_def_id() {
+                    let mut fallback = false;
+
                     if child.ident.name == kw::Underscore {
-                        fallback_map.push((def_id, parent));
-                        return;
+                        fallback = true;
                     }
 
                     if tcx.is_doc_hidden(parent) {
-                        fallback_map.push((def_id, parent));
-                        return;
+                        fallback = true;
                     }
 
                     // If the re-export itself is `#[doc(hidden)]`, deprioritize it.
@@ -519,20 +519,30 @@ pub(in crate::rmeta) fn provide(providers: &mut Providers) {
                         .and_then(|r| r.id())
                         .is_some_and(|id| tcx.is_doc_hidden(id))
                     {
-                        fallback_map.push((def_id, parent));
-                        return;
+                        fallback = true;
                     }
 
                     match visible_parent_map.entry(def_id) {
                         Entry::Occupied(mut entry) => {
-                            // If `child` is defined in crate `cnum`, ensure
-                            // that it is mapped to a parent in `cnum`.
-                            if def_id.is_local() && entry.get().is_local() {
-                                entry.insert(parent);
+                            if !fallback {
+                                // If `child` is defined in crate `cnum`, ensure
+                                // that it is mapped to a parent in `cnum`.
+                                if def_id.is_local() && entry.get().is_local() {
+                                    entry.insert(parent);
+                                }
                             }
                         }
                         Entry::Vacant(entry) => {
-                            entry.insert(parent);
+                            if fallback {
+                                // We do all of the same steps to fallback entries as to
+                                // preferred entries, except for recording them in a separate map.
+                                // It is important to not return early in the fallback cases to
+                                // ensure that we extend the BFS to the children of fallback items.
+                                fallback_map.push((def_id, parent));
+                            } else {
+                                entry.insert(parent);
+                            }
+
                             if child.res.module_like_def_id().is_some() {
                                 bfs_queue.push_back(def_id);
                             }

--- a/tests/ui/suggestions/suggest-path-through-direct-dep-crate/auxiliary/direct-dep-with-nesting.rs
+++ b/tests/ui/suggestions/suggest-path-through-direct-dep-crate/auxiliary/direct-dep-with-nesting.rs
@@ -1,0 +1,14 @@
+#![crate_type = "lib"]
+
+extern crate transitive_dep;
+
+mod private {
+    pub mod inner {
+        pub use crate::transitive_dep::Struct;
+    }
+}
+
+#[doc(hidden)]
+pub mod __private {
+    pub use crate::private::*;
+}

--- a/tests/ui/suggestions/suggest-path-through-direct-dep-crate/auxiliary/direct-dep.rs
+++ b/tests/ui/suggestions/suggest-path-through-direct-dep-crate/auxiliary/direct-dep.rs
@@ -1,0 +1,12 @@
+#![crate_type = "lib"]
+
+extern crate transitive_dep;
+
+mod private {
+    pub use crate::transitive_dep::Struct;
+}
+
+#[doc(hidden)]
+pub mod __private {
+    pub use crate::private::*;
+}

--- a/tests/ui/suggestions/suggest-path-through-direct-dep-crate/auxiliary/transitive-dep.rs
+++ b/tests/ui/suggestions/suggest-path-through-direct-dep-crate/auxiliary/transitive-dep.rs
@@ -1,0 +1,3 @@
+#![crate_type = "lib"]
+
+pub struct Struct;

--- a/tests/ui/suggestions/suggest-path-through-direct-dep-crate/hidden-reexport-of-nested-transitive-dep-item.rs
+++ b/tests/ui/suggestions/suggest-path-through-direct-dep-crate/hidden-reexport-of-nested-transitive-dep-item.rs
@@ -1,0 +1,16 @@
+//@ aux-build: transitive-dep.rs
+//@ aux-build: direct-dep-with-nesting.rs
+
+extern crate direct_dep_with_nesting as direct_dep;
+
+struct Struct;
+//~^ NOTE `Struct` is defined in the current crate
+
+fn main() {
+    let _: direct_dep::__private::inner::Struct = Struct;
+    //~^ ERROR mismatched types
+    //~| NOTE expected `direct_dep::__private::inner::Struct`, found `Struct`
+    //~| NOTE expected due to this
+    //~| NOTE `Struct` and `direct_dep::__private::inner::Struct` have similar names, but are actually distinct types
+    //~| NOTE `direct_dep::__private::inner::Struct` is defined in crate `transitive_dep`
+}

--- a/tests/ui/suggestions/suggest-path-through-direct-dep-crate/hidden-reexport-of-nested-transitive-dep-item.stderr
+++ b/tests/ui/suggestions/suggest-path-through-direct-dep-crate/hidden-reexport-of-nested-transitive-dep-item.stderr
@@ -1,0 +1,23 @@
+error[E0308]: mismatched types
+  --> $DIR/hidden-reexport-of-nested-transitive-dep-item.rs:10:51
+   |
+LL |     let _: direct_dep::__private::inner::Struct = Struct;
+   |            ------------------------------------   ^^^^^^ expected `direct_dep::__private::inner::Struct`, found `Struct`
+   |            |
+   |            expected due to this
+   |
+   = note: `Struct` and `direct_dep::__private::inner::Struct` have similar names, but are actually distinct types
+note: `Struct` is defined in the current crate
+  --> $DIR/hidden-reexport-of-nested-transitive-dep-item.rs:6:1
+   |
+LL | struct Struct;
+   | ^^^^^^^^^^^^^
+note: `direct_dep::__private::inner::Struct` is defined in crate `transitive_dep`
+  --> $DIR/auxiliary/transitive-dep.rs:3:1
+   |
+LL | pub struct Struct;
+   | ^^^^^^^^^^^^^^^^^
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/suggestions/suggest-path-through-direct-dep-crate/hidden-reexport-of-transitive-dep-item.rs
+++ b/tests/ui/suggestions/suggest-path-through-direct-dep-crate/hidden-reexport-of-transitive-dep-item.rs
@@ -1,0 +1,16 @@
+//@ aux-build: transitive-dep.rs
+//@ aux-build: direct-dep.rs
+
+extern crate direct_dep;
+
+struct Struct;
+//~^ NOTE `Struct` is defined in the current crate
+
+fn main() {
+    let _: direct_dep::__private::Struct = Struct;
+    //~^ ERROR mismatched types
+    //~| NOTE expected `direct_dep::__private::Struct`, found `Struct`
+    //~| NOTE expected due to this
+    //~| NOTE `Struct` and `direct_dep::__private::Struct` have similar names, but are actually distinct types
+    //~| NOTE `direct_dep::__private::Struct` is defined in crate `transitive_dep`
+}

--- a/tests/ui/suggestions/suggest-path-through-direct-dep-crate/hidden-reexport-of-transitive-dep-item.stderr
+++ b/tests/ui/suggestions/suggest-path-through-direct-dep-crate/hidden-reexport-of-transitive-dep-item.stderr
@@ -1,0 +1,23 @@
+error[E0308]: mismatched types
+  --> $DIR/hidden-reexport-of-transitive-dep-item.rs:10:44
+   |
+LL |     let _: direct_dep::__private::Struct = Struct;
+   |            -----------------------------   ^^^^^^ expected `direct_dep::__private::Struct`, found `Struct`
+   |            |
+   |            expected due to this
+   |
+   = note: `Struct` and `direct_dep::__private::Struct` have similar names, but are actually distinct types
+note: `Struct` is defined in the current crate
+  --> $DIR/hidden-reexport-of-transitive-dep-item.rs:6:1
+   |
+LL | struct Struct;
+   | ^^^^^^^^^^^^^
+note: `direct_dep::__private::Struct` is defined in crate `transitive_dep`
+  --> $DIR/auxiliary/transitive-dep.rs:3:1
+   |
+LL | pub struct Struct;
+   | ^^^^^^^^^^^^^^^^^
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0308`.

